test_runner/regress/test_tenant_size.py

```diff
@@ -710,3 +710,114 @@ def mask_model_inputs(x):
         return newlist
     else:
         return x
+
+
+@pytest.mark.parametrize("zero_gc", [True, False])
+def test_lsn_lease_size(neon_env_builder: NeonEnvBuilder, test_output_dir: Path, zero_gc: bool):
+    """
+    Compare a LSN lease to a read-only branch for synthetic size calculation.
+    They should have the same effect.
+    """
+
+    conf = {
+        "pitr_interval": "0s" if zero_gc else "3600s",
+        "gc_period": "0s",
+    }
+
+    env = neon_env_builder.init_start(initial_tenant_conf=conf)
+
+    ro_branch_res = insert_and_create_ro_branch(
+        env, env.initial_tenant, env.initial_timeline, test_output_dir
+    )
+
+    tenant, timeline = env.neon_cli.create_tenant(conf=conf)
+    lease_res = insert_and_acquire_lease(env, tenant, timeline, test_output_dir)
+
+    for lhs, rhs in zip(lease_res, ro_branch_res):
+        assert_size_approx_equal(lhs, rhs)
+
+
+def insert_and_acquire_lease(
+    env: NeonEnv, tenant: TenantId, timeline: TimelineId, test_output_dir: Path
+) -> list[int]:
+    sizes = []
+    client = env.pageserver.http_client()
+    with env.endpoints.create_start("main", tenant_id=tenant) as ep:
+        initial_size = client.tenant_size(tenant)
+        log.info(f"initial size: {initial_size}")
+
+        with ep.cursor() as cur:
+            cur.execute(
+                "CREATE TABLE t0 AS SELECT i::bigint n FROM generate_series(0, 1000000) s(i)"
+            )
+        last_flush_lsn = wait_for_last_flush_lsn(env, ep, tenant, timeline)
+        res = client.timeline_lsn_lease(tenant, timeline, last_flush_lsn)
+        log.info(f"result from lsn_lease api: {res}")
+
+        with ep.cursor() as cur:
+            cur.execute(
+                "CREATE TABLE t1 AS SELECT i::bigint n FROM generate_series(0, 1000000) s(i)"
+            )
+            cur.execute(
+                "CREATE TABLE t2 AS SELECT i::bigint n FROM generate_series(0, 1000000) s(i)"
+            )
+            cur.execute(
+                "CREATE TABLE t3 AS SELECT i::bigint n FROM generate_series(0, 1000000) s(i)"
+            )
+
+            cur.execute("CHECKPOINT")
```

Comment:
Why the `CHECKPOINT`?

Reply:
This is an attempt to avoid flakiness when calculating logical size. Taking a checkpoint manually here instead of potentially encountering postgres system scheduled checkpoint. 

Previously we saw that during some runs the size after lease and the size after branching do not match up (not within the `approx_equal` threshold). Not sure if the manual checkpoint will help.

Reply:
Also added explanation as comments.

Reply:
Nice idea.

For full determinism, the recommended way to stop new data from being ingested to PSes is to shut down the endpoints, wait for PSes to catch up with SKs, then shut down SKs.

I think `flush_ep_to_pageserver` does what you want.

Reply:
TIL: note the latest e2e test failure is not caused by this.

Reply:
> e2e test failure is not caused by this PR.
Thanks a lot! I reran the check to pass CI.